### PR TITLE
Fix building with -DFAISS_OPT_LEVEL=avx2

### PR DESCRIPTION
**Pull request by KinglittleQ**
As stated in https://github.com/facebookresearch/faiss/issues/1743, `_swigfaiss.so` will not exist if compiled with `-DFAISS_OPT_LEVEL=avx2`. This diff added a conditional statement before copying it.

### CODE REVIEW

**Comment by mdouze** [comment]
@beauby @h-vetinari any comment on this?

**Review by kj9lj — APPROVED** [review]

File: faiss/python/setup.py [hunk 1]
```diff
@@ -16,17 +16,22 @@
...
 shutil.copyfile("swigfaiss.py", "faiss/swigfaiss.py")
+if os.path.exists("swigfaiss_avx2.py"):
+    shutil.copyfile("swigfaiss_avx2.py", "faiss/swigfaiss_avx2.py")
 
 ext = ".pyd" if platform.system() == 'Windows' else ".so"
 prefix = "Release/" * (platform.system() == 'Windows')
-shutil.copyfile(f"{prefix}_swigfaiss{ext}", f"faiss/_swigfaiss{ext}")
+faiss_exists = os.path.exists(f"{prefix}_swigfaiss{ext}")
+faiss_avx2_exists = os.path.exists(f"{prefix}_swigfaiss_avx2{ext}")
+assert(faiss_exists or faiss_avx2_exists)
```
Comment (h-vetinari) [inline faiss/python/setup.py:26]: This should be raised as an error with an appropriate error message, IMO.

File: faiss/python/setup.py [hunk 2]
```diff
@@ -16,17 +16,22 @@
...
 
-try:
-    shutil.copyfile("swigfaiss_avx2.py", "faiss/swigfaiss_avx2.py")
+if faiss_exists:
+    print("Found swigfaiss")
```
Comment (h-vetinari) [inline faiss/python/setup.py:29]:
```suggestion
    print(f"Copying {prefix}_swigfaiss{ext}")
```

File: faiss/python/setup.py [hunk 3]
```diff
@@ -16,17 +16,22 @@
...
+    shutil.copyfile(f"{prefix}_swigfaiss{ext}", f"faiss/_swigfaiss{ext}")
+
+if faiss_avx2_exists:
+    print("Found swigfaiss_avx2")
```
Comment (h-vetinari) [inline faiss/python/setup.py:33]:
```suggestion
    print(f"Copying {prefix}_swigfaiss_avx2{ext}")
```

**Review by h-vetinari — COMMENTED** [review]
Don't have a strong opinion on this, but left some comments.

Names of the booleans could also be improved, e.g.
`faiss_exists` -> `found_swigfaiss_general`
`faiss_avx2_exists` -> `found_swigfaiss_avx2`

**Comment by KinglittleQ** [comment]
@beauby @h-vetinari Updated.

File: faiss/python/setup.py [hunk 4]
```diff
@@ -16,17 +16,29 @@
...
-    pass
+swigfaiss_general_lib = f"{prefix}_swigfaiss{ext}"
+swigfaiss_avx2_lib = f"{prefix}_swigfaiss_avx2{ext}"
+
+found_swigfaiss_general = os.path.exists(swigfaiss_general_lib)
+found_swigfaiss_avx2 = os.path.exists(swigfaiss_avx2_lib)
+
+assert (found_swigfaiss_general or found_swigfaiss_avx2), \
+        f"Could not find {swigfaiss_general_lib} or " \
+        f"{swigfaiss_avx2_lib}. Faiss may not be compiled yet."
```
Comment (h-vetinari) [inline faiss/python/setup.py:33]:
```suggestion
if not (found_swigfaiss_general or found_swigfaiss_avx2):
    raise FileNotFoundError(f"Could find neither {swigfaiss_general_lib} nor "
                            f"{swigfaiss_avx2_lib}. Faiss may not be compiled yet.")
```

File: faiss/python/setup.py [hunk 5]
```diff
@@ -16,17 +16,29 @@
...
 prefix = "Release/" * (platform.system() == 'Windows')
-shutil.copyfile(f"{prefix}_swigfaiss{ext}", f"faiss/_swigfaiss{ext}")
 
-try:
-    shutil.copyfile("swigfaiss_avx2.py", "faiss/swigfaiss_avx2.py")
-    shutil.copyfile(f"{prefix}_swigfaiss_avx2{ext}", f"faiss/_swigfaiss_avx2{ext}")
-except Exception as e:
-    print(f"Could not move swigfaiss_avx2.py / {prefix}_swigfaiss_avx2{ext} due to:\n{e!r}")
-    pass
+swigfaiss_general_lib = f"{prefix}_swigfaiss{ext}"
```
Comment (beauby) [inline faiss/python/setup.py:25]:
```suggestion
swigfaiss_generic_lib = f"{prefix}_swigfaiss{ext}"
```

File: faiss/python/setup.py [hunk 6]
```diff
@@ -16,17 +16,29 @@
...
+swigfaiss_avx2_lib = f"{prefix}_swigfaiss_avx2{ext}"
+
+found_swigfaiss_general = os.path.exists(swigfaiss_general_lib)
```
Comment (beauby) [inline faiss/python/setup.py:28]:
```suggestion
found_swigfaiss_generic = os.path.exists(swigfaiss_generic_lib)
```

File: faiss/python/setup.py [hunk 7]
```diff
@@ -16,17 +16,29 @@
...
+found_swigfaiss_avx2 = os.path.exists(swigfaiss_avx2_lib)
+
+assert (found_swigfaiss_general or found_swigfaiss_avx2), \
```
Comment (beauby) [inline faiss/python/setup.py:31]:
```suggestion
assert (found_swigfaiss_generic or found_swigfaiss_avx2), \
```

File: faiss/python/setup.py [hunk 8]
```diff
@@ -16,17 +16,29 @@
...
+        f"Could not find {swigfaiss_general_lib} or " \
```
Comment (beauby) [inline faiss/python/setup.py:32]:
```suggestion
        f"Could not find {swigfaiss_generic_lib} or " \
```

File: faiss/python/setup.py [hunk 9]
```diff
@@ -16,17 +16,29 @@
...
+        f"{swigfaiss_avx2_lib}. Faiss may not be compiled yet."
+
+if found_swigfaiss_general:
+    print(f"Copying {swigfaiss_general_lib}")
+    shutil.copyfile(swigfaiss_general_lib, f"faiss/_swigfaiss{ext}")
```
Comment (beauby) [inline faiss/python/setup.py:37]:
```suggestion
if found_swigfaiss_generic:
    print(f"Copying {swigfaiss_generic_lib}")
    shutil.copyfile(swigfaiss_generic_lib, f"faiss/_swigfaiss{ext}")
```

File: faiss/python/setup.py [hunk 10]
```diff
@@ -16,17 +16,29 @@
 shutil.copyfile("__init__.py", "faiss/__init__.py")
 shutil.copyfile("loader.py", "faiss/loader.py")
 shutil.copyfile("swigfaiss.py", "faiss/swigfaiss.py")
```
Comment (beauby) [inline faiss/python/setup.py:18]: Should this be guarded by an `if os.path.exists` as well?

**Review by beauby — COMMENTED** [review]
Thanks for the PR, a few minor suggestions but we can merge it soon.

File: faiss/python/setup.py (hunk 4 above)
Comment (KinglittleQ) [reply faiss/python/setup.py]: Your suggestion is reasonable but I think a simple assertion is enough here.

File: faiss/python/setup.py (hunk 10 above)
Comment (KinglittleQ) [reply faiss/python/setup.py:18]: `swigfaiss.py` is always generated regardless of FAISS_OPT_LEVEL.

File: faiss/python/setup.py [hunk 11]
```diff
@@ -16,17 +16,29 @@
...
+if os.path.exists("swigfaiss_avx2.py"):
+    shutil.copyfile("swigfaiss_avx2.py", "faiss/swigfaiss_avx2.py")
```
Comment (beauby) [inline faiss/python/setup.py:20]: Related to the above: it seems CMake's SWIG target generates `swigfaiss.py` regardless of whether it is an AVX2 build or not (and hence never generates anything called `swigfaiss_avx2.py`). Long term fix will be to introduce separate targets for `libfaiss`/`libfaiss_avx2` (and `swigfaiss`/`swigfaiss_avx2`), but for now I think we can just remove this `shutil.copyfile("swigfaiss_avx2.py", "faiss/swigfaiss_avx2.py")` as it is never executed.

File: faiss/python/setup.py (hunk 4 above)
Comment (h-vetinari) [reply faiss/python/setup.py]: Ultimately up to the maintainers here, but I think bare asserts (outside of tests) are bad style. You should return a specific error.

File: faiss/python/setup.py (hunk 11 above)
Comment (KinglittleQ) [reply faiss/python/setup.py:20]: That's good.

**Pull request closed without merging**